soseki/reader/modeling.py

```diff
@@ -733,70 +781,63 @@ def _compute_answer_accuracy(
     def generate_answer_candidates(
         self,
         input_ids: Tensor,
-        attention_mask: Tensor,
-        token_type_ids: Tensor,
         classifier_logits: Tensor,
         start_logits: Tensor,
         end_logits: Tensor,
-        num_reading_passages: int = 1,
+        num_passages_to_read: int = 1,
         num_answer_candidates_per_passage: int = 1,
         max_answer_length: int = 10,
-    ) -> List[List[AnswerCandidate]]:
-        # check the tensor sizes
-        Q, P, L = input_ids.size()
-        assert input_ids.size() == (Q, P, L)
-        assert attention_mask.size() == (Q, P, L)
-        assert token_type_ids.size() == (Q, P, L)
-        assert classifier_logits.size() == (Q, P)
-        assert start_logits.size() == (Q, P, L)
-        assert end_logits.size() == (Q, P, L)
-
-        # obtain passage indices with highest classifier logits for each question
-        top_classifier_logits, top_passage_idxs = classifier_logits.topk(num_reading_passages, dim=1)
-        assert top_classifier_logits.size() == (Q, min(P, num_reading_passages))
-        assert top_passage_idxs.size() == (Q, min(P, num_reading_passages))
-
-        batch_answer_candidates: List[List[AnswerCandidate]] = []
-        for qi in range(Q):
-            answer_candidates: List[AnswerCandidate] = []
-            for pi, classifier_logit in zip(top_passage_idxs[qi].tolist(), top_classifier_logits[qi].tolist()):
-                answer_spans = self.tokenization._compute_best_answer_spans(
-                    input_ids=input_ids[qi, pi].tolist(),
-                    attention_mask=attention_mask[qi, pi].tolist(),
-                    token_type_ids=token_type_ids[qi, pi].tolist(),
-                    start_logits=start_logits[qi, pi].tolist(),
-                    end_logits=end_logits[qi, pi].tolist(),
-                    num_answer_spans=num_answer_candidates_per_passage,
-                    max_answer_length=max_answer_length,
+    ) -> List[AnswerCandidate]:
```

Comment:
`generate_answer_candidates()` now processes one question at a time while it previously processed a batch of questions.